AtTools/AtEDistortionModel.cxx

```diff
@@ -0,0 +1,191 @@
+#include <Math/Point3D.h>    // for PositionVector3D
+#include <Math/Point3Dfwd.h> // for RhoZPhiPoint, XYZPoint
+#include <Math/Vector2D.h>
+#include <Math/Vector2Dfwd.h> // for XYVector
+#include <Math/Vector3D.h>
+#include <Math/Vector3Dfwd.h> // for XYZVector
+
+#include "TMath.h"
+
+#include <cmath>
+#include <memory> // for allocator
+
+using XYZPoint = ROOT::Math::XYZPoint;
+using XYZVector = ROOT::Math::XYZVector;
+using XYPoint = ROOT::Math::XYPoint;
+using XYVector = ROOT::Math::XYVector;
+
+EFieldMapRef::EFieldMapRef() {}
+
+EFieldMapRef::EFieldMapRef(Int_t rad, Int_t zpos) : _rad(rad), _zpos(zpos) {}
+
+AtEDistortionModel::AtEDistortionModel() : AtSpaceChargeModel() {}
+
+XYZPoint AtEDistortionModel::CorrectSpaceCharge(const XYZPoint &input)
+{
+
+   if (fDistortionMap->size() == 0) {
+      LOG(error) << " Warning! Correction map is empty. No correction applied..."
+                 << "\n";
+      return XYZPoint(input.X(), input.Y(), input.Z());
+   }
+
+   LOG(debug) << " Input point : " << input.X() << " " << input.Y() << " " << input.Z() << "\n";
+
+   Double_t inputRad = TMath::Sqrt(TMath::Power(input.X(), 2) + TMath::Power(input.Y(), 2));
+   Double_t inputAng = TMath::ATan2(input.Y(), input.X());
+   Double_t inputZ = input.Z();
+
+   inputZ = inputZ < 0 ? 0 : inputZ;
+   inputZ = inputZ > 1000 ? 1000 : inputZ;
+
+   auto [zcorr, radcorr, tracorr] = GetCorrectionFactors((Int_t)inputRad, (Int_t)inputZ);
+
+   Double_t outputRad = TMath::Sqrt(TMath::Power(inputRad + radcorr, 2) + TMath::Power(tracorr, 2));
+   Double_t outputAng = inputAng + TMath::ATan2(tracorr, inputRad + radcorr);
+
+   XYZPoint output;
+   output.SetX(outputRad * TMath::Cos(outputAng));
+   output.SetY(outputRad * TMath::Sin(outputAng));
+   output.SetZ(input.Z() - zcorr);
+
+   LOG(debug) << " Output point : " << output.X() << " " << output.Y() << " " << output.Z() << "\n";
+
+   return output;
+}
+
+XYZPoint AtEDistortionModel::ApplySpaceCharge(const XYZPoint &input)
+{
+   XYZPoint p;
+   return p;
+}
+
+Bool_t
+AtEDistortionModel::SetCorrectionMaps(const std::string &zlut, const std::string &radlut, const std::string &tralut)
+{
+
+   fZLUT.open(zlut, std::ifstream::in);
+   fRadLUT.open(radlut, std::ifstream::in);
+   fTraLUT.open(tralut, std::ifstream::in);
+
+   if (!fZLUT.fail() && !fRadLUT.fail() && !fTraLUT.fail()) {
+
+      SetBufferMap(fZLUT, fZMap.get());
+      SetBufferMap(fRadLUT, fRadMap.get());
+      SetBufferMap(fTraLUT, fTraMap.get());
+
+      SetGlobalMap(fZMap.get(), fRadMap.get(), fTraMap.get());
+
+      return kTRUE;
+
+   } else {
+
+      std::cerr << " AtEDistortionModel::SetCorrectionMaps - Error! Missing LUT "
+                << "\n";
+      std::cerr << zlut << "\n";
+      std::cerr << radlut << "\n";
+      std::cerr << tralut << "\n";
+      return kFALSE;
+   }
+
+   return kTRUE;
+}
+
+std::tuple<Double_t, Double_t, Double_t>
+AtEDistortionModel::GetCorrectionFactors(const Int_t &radius, const Int_t &zpos)
```

Comment:
Fine to take references, but is the same as just using an `Int_t radius` (addresses and ints are the same size).

This also doesn't matter, and is just a style thing, but you can just pass around EFieldCorr objects to avoid converting between those and tuples in the code.